swarm/pss/prox_test.go

```diff
@@ -255,93 +243,153 @@ func testProxNetwork(t *testing.T) {
 }
 
 // testRoutine is the main test function, called by Simulation.Run()
-func testRoutine(tstdata *testData, ctx context.Context) error {
-	go handlerChannelListener(tstdata, ctx)
-	go tstdata.sendAllMsgs()
+func testRoutine(d *testData, ctx context.Context) error {
+	go handlerChannelListener(d)
+
+	nodes := make(map[int]*rpc.Client)
+	for i := range d.msgs {
+		nodeClient, err := d.sim.Net.GetNode(d.senders[i]).Client()
+		if err != nil {
+			return err
+		}
+		nodes[i] = nodeClient
+	}
+
+	go d.sendAllMsgs(nodes)
+
+	var res error
+	var done bool
 	received := 0
 
 	// collect incoming messages and terminate with corresponding status when message handler listener ends
+	for !done {
+		select {
+		case <-ctx.Done(): // timeout or cancel
+			res = ctx.Err()
+			done = true
+		case err := <-d.errC:
+			// only first error matters
+			if res == nil {
+				res = err
+			}
+			done = (res != nil)
+		case hn := <-d.msgC:
+			received++
+			log.Debug("msg received", "msgs_received", received, "total_expected", d.requiredMessages, "id", hn.id, "serial", hn.serial)
+		}
+	}
+
+	d.setDone()
+	time.Sleep(time.Millisecond * 16) // allow the nodeMsgHandlers to complete if running
+	close(d.doneC)
+
 	for {
 		select {
-		case err := <-tstdata.errC:
-			return err
-		case hn := <-tstdata.msgC:
+		case <-d.errC:
+			// channel is closed, now function can return
+			return res
+		case <-d.msgC:
 			received++
-			log.Debug("msg received", "msgs_received", received, "total_expected", tstdata.requiredMessages, "id", hn.id, "serial", hn.serial)
-			if received == tstdata.allowedMessages {
-				close(tstdata.doneC)
-				return nil
-			}
 		}
 	}
-	return nil
+	return res
 }
 
-func handlerChannelListener(tstdata *testData, ctx context.Context) {
+func handlerChannelListener(d *testData) {
 	for {
 		select {
-		case <-tstdata.doneC: // graceful exit
-			tstdata.setDone()
-			tstdata.errC <- nil
-			return
-
-		case <-ctx.Done(): // timeout or cancel
-			tstdata.setDone()
-			tstdata.errC <- ctx.Err()
+		case <-d.doneC: // graceful exit
+			close(d.errC)
 			return
 
 		// incoming message from pss message handler
-		case handlerNotification := <-tstdata.handlerC:
+		case handlerNotification := <-d.handlerC:
 			// check if recipient has already received all its messages and notify to fail the test if so
-			aMsgs := tstdata.allowedMsgs[handlerNotification.id]
-			if len(aMsgs) == 0 {
-				tstdata.setDone()
-				tstdata.errC <- fmt.Errorf("too many messages received by recipient %x", handlerNotification.id)
-				return
+			h := handlerNotification.id
+			if len(d.allowedMsgs[h]) == 0 {
+				d.errC <- fmt.Errorf("too many messages received by recipient %x", handlerNotification.id)
+				break
 			}
 
 			// check if message serial is in expected messages for this recipient and notify to fail the test if not
 			idx := -1
-			for i, msg := range aMsgs {
-				if handlerNotification.serial == msg {
+			for i, s := range d.allowedMsgs[h] {
```

Comment:
`d.findAllowedMessage(h, s)` => idx, found? 

but here it would be even enough to have `d.isAllowedMessage(h,s)` if we don't mind searching again.